File: src/services/ollama.ts
import { env } from "../config/env.js";

type OllamaMessage = {
  role: "system" | "user" | "assistant";
  content: string;
};

type OllamaChatResponse = {
  message?: {
    content?: string;
  };
};

type OllamaTagsResponse = {
  models?: Array<{
    name?: string;
    model?: string;
  }>;
};

export async function askOllamaChat(userText: string, options?: { model?: string }) {
  const endpoint = new URL("/api/chat", env.OLLAMA_BASE_URL).toString();
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), env.OLLAMA_TIMEOUT_MS);
  const selectedModel = options?.model?.trim() || env.OLLAMA_CHAT_MODEL;

  const messages: OllamaMessage[] = [
    { role: "system", content: env.OLLAMA_SYSTEM_PROMPT },
    { role: "user", content: userText },
  ];

  try {
    const response = await fetch(endpoint, {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify({
        model: selectedModel,
        stream: false,
        messages,
      }),
      signal: controller.signal,
    });

    const raw = await response.text();
    let parsed: OllamaChatResponse | null = null;
    try {
      parsed = raw ? (JSON.parse(raw) as OllamaChatResponse) : null;
    } catch {
      parsed = null;
    }

    if (!response.ok) {
      throw new Error(`OLLAMA_HTTP_${response.status}: ${raw.slice(0, 300)}`);
    }

    const content = parsed?.message?.content?.trim();
    if (!content) {
      throw new Error("OLLAMA_EMPTY_RESPONSE");
    }

    return {
      text: content,
      model: selectedModel,
      endpoint,
    };
  } finally {
    clearTimeout(timeout);
  }
}

export async function listOllamaModels() {
  const endpoint = new URL("/api/tags", env.OLLAMA_BASE_URL).toString();
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), env.OLLAMA_TIMEOUT_MS);
  try {
    const response = await fetch(endpoint, {
      method: "GET",
      signal: controller.signal,
    });
    const raw = await response.text();
    if (!response.ok) {
      throw new Error(`OLLAMA_TAGS_HTTP_${response.status}: ${raw.slice(0, 300)}`);
    }
    const parsed = raw ? (JSON.parse(raw) as OllamaTagsResponse) : null;
    const models = (parsed?.models ?? [])
      .map((item) => (typeof item.name === "string" ? item.name : typeof item.model === "string" ? item.model : ""))
      .filter((item) => item.length > 0);
    return {
      endpoint,
      models: Array.from(new Set(models)),
    };
  } finally {
    clearTimeout(timeout);
  }
}

File: apps/voice-agent/src/voice_agent/providers/llm/ollama.py
from __future__ import annotations

import aiohttp


class OllamaLLM:
    def __init__(self, base_url: str, model: str) -> None:
        self._base_url = base_url.rstrip("/")
        self._model = model

    async def complete(self, user_text: str, system_prompt: str) -> str:
        payload = {
            "model": self._model,
            "stream": False,
            "messages": [
                {"role": "system", "content": system_prompt},
                {"role": "user", "content": user_text},
            ],
        }

        timeout = aiohttp.ClientTimeout(total=30)
        async with aiohttp.ClientSession(timeout=timeout) as session:
            async with session.post(f"{self._base_url}/api/chat", json=payload) as response:
                if response.status >= 400:
                    raise RuntimeError(f"ollama_failed_{response.status}")
                data = await response.json()
                message = data.get("message") or {}
                return str(message.get("content", "")).strip()
